pkg/providers/version/version.go

```diff
@@ -30,35 +36,94 @@ const (
+	cache      *cache.Cache
+	cm         *pretty.ChangeMonitor
+	kubeClient client.Client
+	httpClient HTTPClient
 }
 
-func NewProvider(kubernetesInterface kubernetes.Interface, cache *cache.Cache) *Provider {
+func NewProvider(cache *cache.Cache, client client.Client, httpClient HTTPClient) *Provider {
 	return &Provider{
-		cm:                  pretty.NewChangeMonitor(),
-		cache:               cache,
-		kubernetesInterface: kubernetesInterface,
+		cm:         pretty.NewChangeMonitor(),
+		cache:      cache,
+		kubeClient: client,
+		httpClient: httpClient,
 	}
 }
 
 func (p *Provider) Get(ctx context.Context) (string, error) {
 	if version, ok := p.cache.Get(kubernetesVersionCacheKey); ok {
 		return version.(string), nil
 	}
-	serverVersion, err := p.kubernetesInterface.Discovery().ServerVersion()
+	version, err := p.GetMinKubernetesVersion(ctx)
 	if err != nil {
 		return "", err
 	}
-	version := fmt.Sprintf("%s.%s", serverVersion.Major, strings.TrimSuffix(serverVersion.Minor, "+"))
 	p.cache.SetDefault(kubernetesVersionCacheKey, version)
 	if p.cm.HasChanged("kubernetes-version", version) {
 		logging.FromContext(ctx).With("version", version).Debugf("discovered kubernetes version")
 	}
 	return version, nil
 }
+
+// GetMinKubernetesVersion ensures that we query all known APIServers for the K8s version.
+// This to handle any scenarios where there may be multiple APIServer reporting different
+// K8s versions.
+func (p *Provider) GetMinKubernetesVersion(ctx context.Context) (string, error) {
+	var endpointSlice v1.EndpointSlice
+	if err := p.kubeClient.Get(ctx, types.NamespacedName{Name: "kubernetes", Namespace: "default"}, &endpointSlice); err != nil {
+		return "", fmt.Errorf("getting endpoints, %w", err)
+	}
+	// This error should never occur
+	minVersion, err := version.ParseGeneric("v99.99.99")
+	if err != nil {
+		return "", fmt.Errorf("parsing version, %w", err)
+	}
+	for _, address := range getAddresses(endpointSlice) {
+		resp, err := p.httpClient.Get(address)
+		if err != nil {
+			return "", fmt.Errorf("sending get request, %w", err)
+		}
+		defer resp.Body.Close()
+		var data map[string]string
+		if err := json.NewDecoder(resp.Body).Decode(&data); err != nil {
+			return "", fmt.Errorf("reading response body, %w", err)
+		}
+		v, err := version.ParseGeneric(data["gitVersion"])
+		if err != nil {
+			return "", fmt.Errorf("parsing kubernetes version, %w", err)
+		}
+		if v.LessThan(minVersion) {
+			minVersion = v
+		}
+	}
+	// Return only the first two version numbers
+	return strings.Join(strings.Split(minVersion.String(), ".")[0:2], "."), nil
+}
+func getAddresses(endpoints v1.EndpointSlice) []string {
+	ports := []string{}
+	// If there are no ports, it's the same as defining all ports.
+	if len(endpoints.Ports) == 0 {
+		ports = []string{""}
+	} else {
+		for _, port := range endpoints.Ports {
+			ports = append(ports, fmt.Sprintf(":%s", strconv.Itoa(int(*port.Port))))
+		}
+	}
+	ret := []string{}
```

Comment:
```suggestion
	var ret []string
```

This is preferred by https://github.com/golang/go/wiki/CodeReviewComments#declaring-empty-slices